libsqlite3-sys/build.rs

```diff
@@ -593,47 +712,377 @@ mod bindings {
+            match api_fn_name {
+                "sqlite3_db_config" => {
+                    // https://sqlite.org/c3ref/c_dbconfig_defensive.html
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_db_config_constchar".to_string(),
+                        generate_varargs_input_idents(field_ident, &bare_fn, &[&const_c_char_type]),
+                    ); // used for SQLITE_DBCONFIG_MAINDBNAME
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_db_config_void_int_mutint".to_string(),
+                        generate_varargs_input_idents(
+                            field_ident,
+                            &bare_fn,
+                            &[&mut_void_type, &c_int_type, &mut_c_int_type],
+                        ),
+                    ); // used for SQLITE_DBCONFIG_LOOKASIDE
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_db_config_int_mutint".to_string(),
+                        generate_varargs_input_idents(
+                            field_ident,
+                            &bare_fn,
+                            &[&c_int_type, &mut_c_int_type],
+                        ),
+                    ); // used for all other configuration verbs
+                }
+                "sqlite3_vtab_config" => {
+                    // https://sqlite.org/c3ref/c_vtab_constraint_support.html
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_vtab_config_int".to_string(),
+                        generate_varargs_input_idents(field_ident, &bare_fn, &[&c_int_type]),
+                    ); // used for SQLITE_VTAB_CONSTRAINT_SUPPORT
+                }
+                _ => {}
+            };
+        }
+
+        let mut wrappers = String::new();
+        for (api_fn_name, api_fn_inputs) in wrapper_fn_inputs_map {
+            let api_fn_ident = syn::Ident::new(&api_fn_name, field_ident.span());
+
+            // get identifiers for each of the inputs to use in the api call
+            let api_fn_input_idents: Vec<syn::Ident> = (&api_fn_inputs)
+                .into_iter()
+                .map(|input| match &input.name {
+                    Some((ident, _)) => ident.to_owned(),
+                    _ => {
+                        panic!("Input has no name {:#?}", input);
+                    }
+                })
+                .collect();
+
+            // generate wrapper and return it as a string
+            let wrapper_tokens = quote! {
+                pub unsafe fn #api_fn_ident(#api_fn_inputs) #api_fn_output {
+                    if sqlite3_api.is_null() {
+                        panic!("sqlite3_api is null");
+                    }
+                    ((*sqlite3_api).#field_ident
+                        .expect(stringify!("sqlite3_api contains null pointer for ", #field_name, " function")))(
+                            #(#api_fn_input_idents),*
+                    )
+                }
```

Comment:
Previously commented on (by me) in https://github.com/rusqlite/rusqlite/pull/631#issuecomment-620874452
"
Extension API function wrappers will be generated for all API functions in the sqlite bindings present at build time. It seems there is a trend towards encouraging building rusqlite against a very recent version of sqlite when any non-ancient features are requested (gating them all with the `modern_sqlite` feature - see #639 ). If this is the case, vtable authors will have little choice but to disable the version check in order to make it possible for end users to use their extension in a variety of environments. This presents a serious safety risk as an attempt to call an API function from a newer version (i.e. from the wrapper at

https://github.com/rusqlite/rusqlite/blob/cf658af142399be749bacb0387d54ce836b85f1a/libsqlite3-sys/build.rs#L638

) would result in reading past the end of that struct into memory we don't own, and then executing what we find at the target of a pointer we find there.

If we are going to encourage people to disable the version check by making it difficult to build against a minimum required version of sqlite, then we should probably protect the wrappers by including a runtime version check in them (probably using Once) and refusing to attempt to execute any belonging to a version newer than the runtime environment supports.
"

And https://github.com/rusqlite/rusqlite/pull/631#issuecomment-620877082
"
Version requirements for the API routines are documented in comments in `sqlite3ext.h` (e.g. https://github.com/rusqlite/rusqlite/blob/master/libsqlite3-sys/sqlite3/sqlite3ext.h#L287) but I am not sure what the best way would be to capture this and pass it through to the wrapper function generator so we can populate a required version test in each of them. 
"

Reply:
Runtime version checking has now been implemented by generating a separate set of rust-bindgen bindings from `sqlite3ext.h` including comments and parsing the `sqlite_api_routine` fields along with the inline version comments that indicate minimum versions for each section of added fields.